### PR TITLE
Fix: App crashes when tapping more and share button in the Reader Details screen

### LINKED ISSUE #20487: Reader: Crash on iPad after tapping more/share button in the details view
Crash on iPad after tapping more button in the details view

## Steps to reproduce

1. Launch Jetpack App on iPad
2. Select the Reader tab
3. Select any post
4. Click the top right ellipsis "more" button
5. Observe crash

Reproducible on iPad, Jetpack 22.0

## Sentry Logs

Sentry Issue: [JETPACK-IOS-164](https://a8c.sentry.io/issues/4058805512/?referrer=github_integration)

```
NSInternalInconsistencyException: Tried to set <UIBarButtonItem: 0x13b3b2000> target=0x138b3e600 action=didTapMenuButton: image=<UIImage:0x280cddc20 named(main: icon-menu-vertical-ellipsis) {4, 18} renderingMode=alwaysTemplate> as sourceView of <_UIAlertControllerActionSheetRegularPresentationController: 0x1373beb10>, even though it is not a kind of UIView.
  File "ReaderShowMenuAction.swift", line 197, in ReaderShowMenuAction.execute
  File "ReaderMenuAction.swift", line 18, in ReaderMenuAction.execute
  File "ReaderDetailCoordinator.swift", line 413, in ReaderDetailCoordinator.showMenu
  File "ReaderDetailCoordinator.swift", line 684, in ReaderDetailCoordinator.didTapMenuButton
  File "ReaderDetailViewController.swift", line 712, in ReaderDetailViewController.didTapMenuButton
...
(26 additional frame(s) were not displayed)
```

## Video

https://user-images.githubusercontent.com/4062343/230423668-827755f4-c53a-4a2b-ad32-deead1b45b9b.mp4

### PR DESCRIPTION
**Pull request by salimbraksa**
Fixes #20487

## Description

This PR fixed a bug that crashed the app when the user taps the more or share button in the Reader Details screen. This bug is only affecting iPad users.

## Root Cause

This is the exact change in https://github.com/wordpress-mobile/WordPress-iOS/pull/20363 that caused the crash.

```diff
    func barButtonItem(with image: UIImage, action: Selector) -> UIBarButtonItem {
        let image = image.withRenderingMode(.alwaysTemplate)
-      let button = UIButton(frame: CGRect(x: 0, y: 0, width: 44.0, height: image.size.height))
-      button.setImage(image, for: UIControl.State())
-      button.addTarget(self, action: action, for: .touchUpInside)
- 
-       return UIBarButtonItem(customView: button)
+      return UIBarButtonItem(image: image, style: .plain, target: self, action: action)
    }
```

The code above removes usage of `UIButton` when creating a `UIBarButtonItem`, so the `action` param is now passed to a `UIBarButtonItem` instead of a `UIButton`. But the `sender` param of the action handlers was not updated accordingly.

```swift
// This method can be found in "ReaderDetailViewController.swift".
@objc private func didTapMenuButton(_ sender: UIButton) {
   // ...
}
```
The type of `sender` param type should be `UIBarButtonItem` instead of `UIButton`. The compiler didn't catch this, and I didn't either.

At runtime, however, the `sender` param type was `UIBarButtonItem`, which passed through several methods until it reached the following code.

```swift
// Can be found in `ReaderShowMenuAction.swift` and `PostSharingController.swift`.
if let presentationController = alertController.popoverPresentationController {
    presentationController.permittedArrowDirections = .any
    presentationController.sourceView = anchor
    presentationController.sourceRect = anchor.bounds
}
```
The `presentationController.sourceView`'s type is `UIView` and `anchor`'s type at runtime is `UIBarButtonItem`, thus the crash.

## Screenshots

| More Button | Share Button |
| ------------ | -------------- |
|  |  |

## Test Instructions

Please perform the following tests on iPhone and iPad.

1. Install the Jetpack app and log in.
2. Head to **Reader > Following** or **Reader > Discover**
3. Tap the More Button on any post.
4. **Expect** a sheet to show up on iPhone and a popover to show up on iPad.
5. Tap the post itself to reach the Reader Detail screen.
6. Tap the More button.
7. **Expect** a sheet/popover to appear.
8. Tap the Share button.
9. **Expect** a sheet/popover to appear.

## Regression Notes
1. Potential unintended areas of impact
Double check the More and Share buttons work as expected in Reader Details and Reader Following / Discover screens.

| | Reader Details | Reader Following & Discover Feeds |
| - | -------------- | ----------------------------------- |
| iPad | ✅ | ✅ |
| iPhone | ✅ | ✅ |

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing.

3. What automated tests I added (or what prevented me from doing so)
None.

## PR submission checklist:
- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Review by guarani — APPROVED** [review]
Tested and works well! Thanks for the fix, @salimbraksa!

**Comment by mokagio** [comment]
I find it spooky that this PR keeps failing in `AccoutSettingsServiceTests` however I don't see how the changes here could have affected those tests.

Moreover, there might be something dodgy down in the reporting layer, because [looking at the logs](https://buildkite.com/automattic/wordpress-ios/builds/13637#018772eb-257c-413e-876e-6d3e84f7628b/534-641) we can see that the test eventually succeeds after retrying automatically

image

I see no reason to hold merging this PR because of that.

**Comment by tiagomar** [comment]
Hey @mokagio @salimbraksa , `testUpdateFailure()` is [the least reliable unit test we have](https://buildkite.com/organizations/automattic/analytics/suites/wordpress-ios/tests?branch=trunk&filter=reliability&period=28days) but in this case we have a 100% reliable test failing which is `testShowShareSheet()` and, for some unknown reason, it doesn't appear in the console logs with the red cross. However, we can see it listed among the Failing tests. So we'll need to either fix something in the code or fix the test.

```
Failing tests:
  AccountSettingsServiceTests.testUpdateFailure()
  AccountSettingsServiceTests.testUpdateFailure()
  ReaderDetailCoordinatorTests.testShowShareSheet()
```

image

image

File: RELEASE-NOTES.txt [hunk 1]
```diff
@@ -8,6 +8,7 @@
 * [***] [internal] Refactor uploading photos (from the device photo, the Free Photo library, and other sources) to the WordPress Media Library. Affected areas are where you can choose a photo and upload, including the "Media" screen, adding images to a post, updating site icon, etc. [#20322]
 * [**] [WordPress-only] Warns user about sites with only individual plugins not supporting core app features and offers the option to switch to the Jetpack app. [#20408]
 * [**] Add a "Personalize Home Tab" button to the bottom of the Home tab that allows changing cards visibility. [#20369]
+* [*] [Reader] Fix an issue that was causing the app to crash when tapping the More or Share buttons in Reader Detail screen. [#20490]
```
Comment (mokagio) [inline RELEASE-NOTES.txt:11]: Notice that since this landed in the release branch after code freeze, the App Store release notes won't include it.

**Pull request merged**